### PR TITLE
[HZ-1007] Offload MapStore/MapLoader Operations

### PR DESCRIPTION
**Pull request by ahmetmircik**
ee counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/5092

For design and benchmarks please see [design doc](https://github.com/ahmetmircik/hazelcast/blob/fix/5.2/opSteps/docs/design/map/map-store-operation-offloading.md)

**Modifications:**
- MapStore/MapLoader operations are offloaded from partition threads so they don't block other operations queued behind.
- Introduced `Step`, `StepSupplier` and `StepRunner` as an underlying engine to realize offloading mechanism
- All map-store interacting operations are modeled as a series of `Step`s. These steps only works when there is a map-store configured otherwise i keep all flow as in now.

closes https://github.com/hazelcast/hazelcast/issues/6646, https://github.com/hazelcast/hazelcast/issues/9467, https://github.com/hazelcast/hazelcast/issues/14483, https://github.com/hazelcast/hazelcast/issues/11237

Depends on config changes in this PR: https://github.com/hazelcast/hazelcast/pull/21658

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-ee-tests

**Comment by ahmetmircik** [comment]
run-ee-tests

**Comment by kwart** [comment]
run-lab-run

**Comment by ahmetmircik** [comment]
run-ee-tests

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:1531]: This is simple LinkedHashMap Is this accessed from a single thread only? If so we should document that.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java:50]: How long does a typical get operation take when the entry is present in memory?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java:98]: Why is this one not volatile?

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java:192]: I am not familiar with the migration subsystem - does this also run on partition threads?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java:119]: What does tiered store has to do with enabling / disabling offload?

File: docs/design/map/map-store-operation-offloading.md
Comment (frant-hartm) [inline docs/design/map/map-store-operation-offloading.md:132]: We now have 2 implementations for map operations that are offloadable - the original which is used when offload is false, or when there is no map store, and the new implementation which is used when we have a map store and offload set to true (default).

- changes to the operations will need to be done in 2 places, this will inevitably lead to bugs when we forget to update one place or another.
- the offload implementation is not tested as much as the one without offload because most tests don't use mapstore, this makes it rather risky.

I see this as the main issue with the PR.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

**Comment by frant-hartm** [comment]
@vbekiaris This certainly needs one or two more reviewers.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/GetOpSteps.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/GetOpSteps.java:62]: same guarantees exist with its previous working, 2nd `READ` will see the loaded value and will return it, there is no load for subsequent `READ` if entry already exists in memory.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java]: fixed by putting all checks in `StepSupplier`, it is a central place to control all.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:131]: Removals need  ~O(1) execution time, for this reason i use Set.

File: hazelcast/src/test/java/com/hazelcast/json/MapIndexJsonTest.java
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/json/MapIndexJsonTest.java]: i don't like metrics polluting logs when a test fails locally. But this is a left-over, reverted it.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/StaticParams.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/StaticParams.java:109]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java]: fixed with volatile.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java:192]: yes, this part runs on partition threads.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java:119]: it has its own thread marking before and after runs and also it will not be permitted to run both map-store and tiered-store on the same map.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:133]: Yes good points.

>changes to the operations will need to be done in 2 places, this will inevitably lead to bugs when we forget to update one place or another.

Mostly using same methods from DefaultRecordStore implementation, but there are duplications of flow, this needs to be addressed, I left this as a second task to future because i want a safe switch to previous working mechanism if there appears some unexpected issues. 

>the offload implementation is not tested as much as the one without offload because most tests don't use mapstore, this makes it rather risky.

This current PR makes use of this new offload mechanism to run all tests independent of they have map-store configured or not and all os and ee tests are passing(nightly ones included). I will create a new jenkins job to run periodically to catch possible inconsistencies, this job will run in same manner with this PR.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:155]: This is only for metrics system, it can read counts from other threads than partitioned ones.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:208]: ok, will do and add this to design doc.

Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:209]: done.

File: hazelcast/src/test/java/com/hazelcast/internal/diagnostics/SlowOperationPluginTest.java
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/internal/diagnostics/SlowOperationPluginTest.java:69]: this will be fixed after merging `MapStoreConfig#offload` PR.

File: hazelcast/src/test/java/com/hazelcast/map/EntryProcessorTest.java
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/EntryProcessorTest.java]: fixed.

**Comment by ahmetmircik** [comment]
run-nightly-tests

**Comment by ahmetmircik** [comment]
run-ee-tests

**Comment by ahmetmircik** [comment]
run-nightly-tests

**Comment by ahmetmircik** [comment]
run-ee-tests

**Comment by ahmetmircik** [comment]
run-nightly-tests

**Comment by ahmetmircik** [comment]
run-ee-tests

File: docs/design/map/map-store-operation-offloading.md
Comment (frant-hartm) [reply docs/design/map/map-store-operation-offloading.md:209]: From the charts it looks like the new implementation has slightly lower latency ~400 us vs ~450 us for master and higher throughput. I would expect them to be closer.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
Comment (frant-hartm) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:131]: But aren't the items removed in the order they are added?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java
Comment (frant-hartm) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java:123]: >  it will not be permitted to run both map-store and tiered-store on the same map.
why not? it doesn't make sense.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java:123]: I was referring this discussion https://hazelcast.slack.com/archives/C01PD4KN9FB/p1635335337046200, but seems my interpretation of it can be not true since there seems no final decision yet. But let's address this separately in another PR.

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java:131]: In some cases, like migration, there can more than 1 call to remove operation, Set usage makes things easier in those cases also.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:209]: When there is no db, all phases of operation are executed in partition thread, the diff from current code is no db related code is executed.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapService.java
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapService.java:112]: Unused constant

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java:197]: Could you document how we handle exception which can happen in the step transitions or while running the operation in the offloading thread? Also, what is expected from the operation steps to avoid corrupting the internal state of the system in case of the exceptions.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/BaseRemoveOperation.java
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/BaseRemoveOperation.java:62]: Currently, we're performing this `this.dataOldValue = mapServiceContext.toData(state.getOldValue());` serialization call twice both in the child classes and here. This situation also applies to the `BasePutOperation` and its child classes.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java:52]: The fact that this class has too many fields gives me the feeling that it will create a litter. But especially the benchmark result, which Frantisek requested, shows that currently this doesn't have a significant effect. I'm writing to note if the bottleneck slides here later -> keeping these state fields in a Map<String, Object> can reduce the memory usage stems from by these fields, most of which are optional.

File: docs/design/map/map-store-operation-offloading.md
Comment (ufukyilmaz) [inline docs/design/map/map-store-operation-offloading.md:195]: Could you also note the relationship of this offloaded operations with the operations that do not interact with the map store? e.g. `map.getEntryView(K key)`

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java:234]: This name can be confused with the other `StepResponseHandler`. `OffloadedStepResponseHandler` would be better.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java]: done.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:195]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/State.java:52]: We will lose type checks in that case if we make use of `Map<String, Object>`.
Now in this simple form it works as intended, i will keep it as is for now.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/BaseRemoveOperation.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/BaseRemoveOperation.java]: good catch, fixed, it should be one time conversion now.

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java]: Added docs for `StepSupplier#nextStep`

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapService.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapService.java]: removed.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md]: Yes, it has same guarantee with current behavior for sync `put`. But only `putAsync` can be not-observed since we have currently an implicit order there for the same key.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:115]: For queued operations, no new object is created, i'm using same object reference while queueing.
Current operation queues can also grow infinitely, to avoid OOME, back-pressure should be enabled.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:70]: It is same with default executor currently in use, max worker count is 16 and queue capacity is int-max. Burden is on user to find a better setting for the use case.
Regarding blocking case, yes it is possible and this is better than the current situation since partition threads are not blocked.

File: docs/design/map/map-store-operation-offloading.md
Comment (ufukyilmaz) [reply docs/design/map/map-store-operation-offloading.md:115]: I also think that a new backpressure mechanism that specifically applies to these per-map queued offloaded operations would be nice improvement. Our current backpressure regulator has some deadlock issues (see: https://github.com/hazelcast/hazelcast/issues/18778#issuecomment-1194095299) since it also applies blocking on the internal operations, therefore it's not applicable for many production env.

Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:115]: Better to address those issues in a different PR. 
Reason is: Back-pressure mechanism must be a general implementation and must be at a different level than data-structures. The issues of it is out of scope of the work in this PR. In this PR, we only change the place of queued operations by moving them from operation-queues to offloaded queues, before they are waiting end of map-store interaction in operation queues, now they are waiting it in offloaded-operation-queues,  i expect memory consumptions are identical with or without this PR.

File: docs/design/map/map-store-operation-offloading.md
Comment (ufukyilmaz) [reply docs/design/map/map-store-operation-offloading.md:47]: This default `true` value for this offloading flag is a more preferable value for new users, but for old users it does not offer a safe switch for version upgrade. I think we should discuss this tradeoff.

**Comment by ufukyilmaz** [comment]
run-nightly-tests

**Comment by ahmetmircik** [comment]
run-nightly-tests

**Comment by ahmetmircik** [comment]
run-ee-tests

File: docs/design/map/map-store-operation-offloading.md
Comment (vbekiaris) [reply docs/design/map/map-store-operation-offloading.md:205]: Good, can you please clarify this in the text above? How do we ensure that the results of a `map.put` are visible in `getEntryView` if the former is queued but the latter is not?

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:75]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseHandler.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseHandler.java:36]: renamed as `StepResponseUtil`

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepAwareOperation.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepAwareOperation.java:24]: I left it for a future time after current api becomes a bit more mature, i would like to address it in a different PR focused for the point. Would you mind to create an issue for this?

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/MapMigrationAwareService.java:220]: This copying is needed to escape from `ConcurrentModificationException`.
Inside `sendResponse`, we also remove operation from Set.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md:47]: >old users it does not offer a safe switch

@ufukyilmaz can you please comment on this, why do you think it cannot be a safe switch/

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java:123]: Further info: I had a look at current code base and tiered-store feature is coded as it expects no map-store interaction. So for now it is safe to assume there will be no map-store when tiered-store is enabled.

File: docs/design/map/map-store-operation-offloading.md
Comment (ufukyilmaz) [reply docs/design/map/map-store-operation-offloading.md:47]: MapLoader/MapStore calls that were not previously offloaded will start to be offloaded after the version upgrade since the default value is enabled.

File: docs/design/map/map-store-operation-offloading.md
Comment (ahmetmircik) [reply docs/design/map/map-store-operation-offloading.md]: done

File: hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/writebehind/TransactionsWithWriteBehind_whenNoCoalescingQueueIsFullTest.java
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/writebehind/TransactionsWithWriteBehind_whenNoCoalescingQueueIsFullTest.java:280]: why extend waiting in this test?

File: hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreTest.java
Comment (vbekiaris) [inline hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreTest.java:362]: `clear` is already tested in `testMapstoreDeleteOnClear` test; if the idea is to test the same with an `EAGER` map store, maybe reflect it in the test name?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseUtil.java
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseUtil.java:35]: what about this todo item?

**Review by vbekiaris — APPROVED** [review]
The design looks ok. I added some minor comments. As discussed in comments, we should definitely setup a nightly job to execute `IMap` tests, forcing usage of the offloaded steps engine.
I feel it would be cleaner to setup the offloaded steps engine within `operationservice` package (eg probably offloaded anti-entropy implementation could be simplified, since it performs some steps on partition threads and some on async executor, so there is a strong indication that the op-steps engine could be more broadly useful). Let's address this in a subsequent PR.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java]: done

File: hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/writebehind/TransactionsWithWriteBehind_whenNoCoalescingQueueIsFullTest.java
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/writebehind/TransactionsWithWriteBehind_whenNoCoalescingQueueIsFullTest.java]: it is left-over, restored old timing.

File: hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreTest.java
Comment (ahmetmircik) [reply hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/MapStoreTest.java]: it is left-over, deleted it.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseUtil.java
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepResponseUtil.java]: it is left-over, deleted it.

**Review by frant-hartm — APPROVED** [review]

**Pull request merged**
